Add wrapper for clip_if() and children_if()

jradha11 opened the pull request.

KarthikRIyer commented on src/test/java/io/opentimeline/TimelineTest.java:
```diff
@@ -247,4 +247,41 @@ public void testTracks() throws ChildAlreadyParentedException {
+            assertTrue((result.get(i)).isEquivalentTo(composableChildrenList.get(i)));
+        }
+
+        //test for null pointer exception when search_range is null
+        assertThrows(NullPointerException.class,
+                ()->{timeline.childrenIf(null, false);});
```
Split this into a separate test. Something like `testChildrenIfNullTimeRange`. And use the Test annotation like this:

`@Test(expected=NullPointerException.class)`

jradha11 replied:
`@Test(expected=NullPointerException.class)` is used in JUnit4. As we are using JUnit5, simple `assertThrows` would serve the purpose inside the test block.

jradha11 replied:
Have seggregated the tests.

KarthikRIyer commented on src/test/java/io/opentimeline/TimelineTest.java:
```diff
@@ -247,4 +247,41 @@ public void testTracks() throws ChildAlreadyParentedException {
     }
+
+    @Test
+    public void testChildrenIf() throws ChildAlreadyParentedException{
```
Rename to something like `testChildrenIfEquality`

jradha11 replied:
Done

KarthikRIyer commented on src/test/java/io/opentimeline/TimelineTest.java:
```diff
@@ -247,4 +247,41 @@ public void testTracks() throws ChildAlreadyParentedException {
+        TimeRange search_range = new TimeRange(
+                new RationalTime(0, 1),
+                new RationalTime(40, 1));
+        List<Composable> result = timeline.childrenIf(search_range, false);
```
Add children (like clips, gaps, effects) to the Track and test those.

jradha11 replied:
Implemented

KarthikRIyer commented on src/test/java/io/opentimeline/TimelineTest.java:
```diff
@@ -247,4 +247,41 @@ public void testTracks() throws ChildAlreadyParentedException {
+        assertThrows(NullPointerException.class,
+                ()->{timeline.childrenIf(null, false);});
+
+        try{
+            timeline.close();
+            V1.close();
+            V2.close();
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
```
Instead of closing these three in a separate try-catch block, use try with resources. See [this](https://stackoverflow.com/a/47175609) SO answer for reference. 

All OTIO objects implement the `AutoCloseable` interface. See [this](https://docs.oracle.com/javase/8/docs/api/java/lang/AutoCloseable.html) for reference. The gist is, when you use an autocloseable with `try with resources`, the `close` method is automatically called when the control exits the try block. We implement the `close` method and free any allocated resources (here, the native memory allocated) or close network connections, etc.

jradha11 replied:
added tests using `try-with-resources`

KarthikRIyer commented on src/main/java/io/opentimeline/opentimelineio/Timeline.java:
```diff
@@ -206,7 +206,7 @@ public Timeline build() {
     }
 
-    private native Composable[] childrenIfNative(TimeRange search_range, boolean shallow_search);
+    private native <T extends Composable> T[] childrenIfNative(TimeRange search_range, boolean shallow_search);
```
You need to be able to specify a `descendedFrom` parameter. Take a look at how it's impleented for `eachChild` [here](https://github.com/OpenTimelineIO/OpenTimelineIO-Java-Bindings/blob/cdf969d5c354a0e00e30d2b5272e7073848a4296/src/main/java/io/opentimeline/opentimelineio/Timeline.java#L146).

jradha11 replied:
Implemented

KarthikRIyer commented on src/test/java/io/opentimeline/TimelineTest.java:
```diff
@@ -249,39 +249,64 @@ public void testTracks() throws ChildAlreadyParentedException {
+            assertTrue(stack.appendChild(V1));
+            assertTrue(stack.appendChild(V2));
+            timeline.setTracks(stack);
+            List<Composable> composableChildrenList = Arrays.asList(V1,C1,V2,C2);
+            TimeRange search_range = new TimeRange(
+                    new RationalTime(0, 1),
+                    new RationalTime(40, 1));
+            List<Composable> result = timeline.childrenIf(search_range, false);
+            for(int i = 0; i < composableChildrenList.size(); i++){
+                assertTrue((result.get(i)).isEquivalentTo(composableChildrenList.get(i)));
+            }
```
Once you implement the `descendedFrom` parameter you should test getting only the clips.

jradha11 replied:
implemented

KarthikRIyer commented:
@jradha11 This looks good. You might have missed a few Composable types like Stack. Could you please check if you've covered all of them?

Also, I would add a check in the tests for the size equality of the reference and result Lists.

jradha11 commented:
@KarthikRIyer 
Nice point about the test. I will check for length too in the equality cases. 
Also, I will be adding the remaining composable types - `Item`, `Composition`, `Transition`, and `Stack`.

KarthikRIyer commented:
Looks good!

reinecke commented on src/main/cpp/io_opentimeline_opentimelineio_Timeline.cpp:
```diff
@@ -187,3 +189,74 @@ Java_io_opentimeline_opentimelineio_Timeline_getVideoTracksNative(
+    auto thisHandle =
+            getHandle<SerializableObject::Retainer<Timeline>>(env, thisObj);
+    auto timeline = thisHandle->value;
+    optional<TimeRange> searchRange = nullopt;
+    searchRange = timeRangeFromJObject(env, searchRangeTimeRange);
```
Could these two lines be combined since the `nullopt` value in `searchRange`  overwritten on the next line?

reinecke commented on src/main/cpp/io_opentimeline_opentimelineio_Timeline.cpp:
```diff
@@ -187,3 +189,74 @@ Java_io_opentimeline_opentimelineio_Timeline_getVideoTracksNative(
+        auto result = timeline->children_if<Track>(&errorStatus, searchRange, shallowSearch);
+        processOTIOErrorStatus(env, errorStatus);
+        env->ReleaseStringUTFChars(name, str);
+        return trackRetainerVectorToArray(env, *(new std::vector<SerializableObject::Retainer<Track>>(result)));
+    }
+    else if(strcmp(str,"io.opentimeline.opentimelineio.Gap") == 0){
+        auto result = timeline->children_if<Gap>(&errorStatus, searchRange, shallowSearch);
+        processOTIOErrorStatus(env, errorStatus);
+        env->ReleaseStringUTFChars(name, str);
+        return gapRetainerVectorToArray(env, *(new std::vector<SerializableObject::Retainer<Gap>>(result)));
+    }
+    else if(strcmp(str,"io.opentimeline.opentimelineio.Stack") == 0){
+        auto result = timeline->children_if<Stack>(&errorStatus, searchRange, shallowSearch);
+        processOTIOErrorStatus(env, errorStatus);
+        env->ReleaseStringUTFChars(name, str);
+        return stackRetainerVectorToArray(env, *(new std::vector<SerializableObject::Retainer<Stack>>(result)));
+    }
+    else if(strcmp(str,"io.opentimeline.opentimelineio.Transition") == 0){
+        auto result = timeline->children_if<Transition>(&errorStatus, searchRange, shallowSearch);
+        processOTIOErrorStatus(env, errorStatus);
+        env->ReleaseStringUTFChars(name, str);
+        return transitionRetainerVectorToArray(env, *(new std::vector<SerializableObject::Retainer<Transition>>(result)));
+    }
+    else if(strcmp(str,"io.opentimeline.opentimelineio.Composition") == 0){
+        auto result = timeline->children_if<Composition>(&errorStatus, searchRange, shallowSearch);
+        processOTIOErrorStatus(env, errorStatus);
+        env->ReleaseStringUTFChars(name, str);
+        return compositionRetainerVectorToArray(env, *(new std::vector<SerializableObject::Retainer<Composition>>(result)));
+    }
+    else if(strcmp(str,"io.opentimeline.opentimelineio.Item") == 0){
+        auto result = timeline->children_if<Item>(&errorStatus, searchRange, shallowSearch);
+        processOTIOErrorStatus(env, errorStatus);
+        env->ReleaseStringUTFChars(name, str);
+        return itemRetainerVectorToArray(env, *(new std::vector<SerializableObject::Retainer<Item>>(result)));
+    }
+    //default return for Composable Type
+    auto result = timeline->children_if(&errorStatus, searchRange, shallowSearch);
+    processOTIOErrorStatus(env, errorStatus);
+    env->ReleaseStringUTFChars(name, str);
+    return composableRetainerVectorToArray(env, *(new std::vector<SerializableObject::Retainer<Composable>>(result)));
```
Looking at all these conditional branches, the only thing that varies between the cases is the return value, could all the code that is the same between cases be pulled up out of the conditional to remove the code duplication?

For example:
```
auto result = timeline->children_if<Clip>(&errorStatus, searchRange, shallowSearch);
processOTIOErrorStatus(env, errorStatus);
env->ReleaseStringUTFChars(name, str);
        
if(strcmp(str,"io.opentimeline.opentimelineio.Clip") == 0){
    return clipRetainerVectorToArray(env, *(new std::vector<SerializableObject::Retainer<Clip>>(result)));
}
else if(strcmp(str,"io.opentimeline.opentimelineio.Track") == 0){
    return trackRetainerVectorToArray(env, *(new std::vector<SerializableObject::Retainer<Track>>(result)));
}
...The rest of the cases
```

KarthikRIyer replied:
@reinecke the template for children_if varies too.   I don't think we can pull out those statements. I don't like this solution too. I'm trying to see if there's something more elegant, but I haven't found anything yet.

jradha11 replied:
@KarthikRIyer I'm working on an inline util to return the corresponding composable jObjectArray for `children_if`.  This will reduce code repetition.

jradha11 replied:
Also, we can remove the `processOTIOErrorStatus(env, errorStatus);` out of the conditional statements but `env->ReleaseStringUTFChars(name, str);` need to be executed before every return statement to free the memory allocated to `str`. If not executed then it'll lead to memory leaks which is causing the timeline tests to fail.

jradha11 replied:
@reinecke added generic util that can be used across classes using `children_if`

KarthikRIyer replied:
In your template util function, if you declare a jObjectArray before your if-else block, and assign the result to that variable, you should be able to pull out the string release function and put it after the if-else block.

jObjectArray is after all just a `_jobject` pointer, so this should work.
See the Reference Types section in the [JNI docs](https://docs.oracle.com/javase/8/docs/technotes/guides/jni/spec/types.html#primitive_types).

KarthikRIyer replied:
Also, @reinecke @jradha11 what do you think about a dispatch table pattern instead of `if-else`?
Something like std::map<std::string, std::function>?

@jradha11 if you are unfamiliar with this pattern, we use this in the [SVG adapter](https://github.com/PixarAnimationStudios/OpenTimelineIO/blob/6ce63ff7bdb5b90f66aa090d5689f566ca7b5b33/src/py-opentimelineio/opentimelineio/adapters/svg.py#L538) (it's in python but you'll get the gist).

jradha11 replied:
We can use a map to get the corresponding function for a `clsNameString`. Sounds about good to me. I'll see how it can be implemented here.

reinecke replied:
I totally missed the `children_if` template parameter! In that case, if there is a way to factor out what's common, great, otherwise we do the best we can.

jradha11 replied:
Implemented the dispatch table.

reinecke commented on src/main/cpp/io_opentimeline_opentimelineio_Timeline.cpp:
```diff
@@ -187,3 +189,74 @@ Java_io_opentimeline_opentimelineio_Timeline_getVideoTracksNative(
+    auto timeline = thisHandle->value;
+    optional<TimeRange> searchRange = nullopt;
+    searchRange = timeRangeFromJObject(env, searchRangeTimeRange);
+    jclass cls = env->GetObjectClass(descendedFromClass);
+    jmethodID getName = env->GetMethodID(cls, "getName", "()Ljava/lang/String;");
+    auto name = (jstring)env->CallObjectMethod(descendedFromClass, getName);
+    const char* str = env->GetStringUTFChars(name, NULL);
```
Consider giving these variables more descriptive names to help code readability. It can be difficult to understand the intended contents of a string named `str`.

jradha11 replied:
Sure, I will change with descriptive variable name.

KarthikRIyer replied:
I think this kind of bad naming is present throughout the JNI layer 😅

Maybe this change could be a different PR?

reinecke replied:
Oh, hah. If we're doing this same operation in a lot of places, it could be useful to create a function to abstract it. Yeah, if this is following some convention in the codebase, it's probably best to address it all at once in a separate PR. I prefer consistency in that case :)

jradha11 replied:
I have tried to keep the naming comprehensible. As @KarthikRIyer said, we can improve on the conventions in a new PR. Shall I resolve it for now?

KarthikRIyer replied:
yes

reinecke reviewed (changes requested):
Looking good! I had a few notes to help code readability and maintainability, but nothing major. Glad to have someone keeping the API up to date!

jradha11 commented:
@reinecke @KarthikRIyer 

I had an observation regarding CI check. In the last test that I added `testChildrenIfComposableEquality` in `CompositionTest.java` ( commit: b00d62d) , I got an exception: `UnsupportedOperationException: method not implemented for this class` as the verbose output ( that was because `range_of_all_children` was not implemented in `composition.cpp`). But still, there was no build error in spite of the exception in tests. I am attaching the screenshot of the verbose output. 

![Screenshot from 2021-09-03 16-29-36](https://user-images.githubusercontent.com/59576984/131995215-cd696f9c-27fe-417b-a552-9365a0e2b60c.png)
> error output

![Screenshot from 2021-09-03 16-29-45](https://user-images.githubusercontent.com/59576984/131995282-29535dad-86bf-41a0-9bf8-300f029a9410.png)
> successful build

I think we need to look into why CI didn't fail.

reinecke commented:
Yikes! Good catch!

jradha11 commented:
@KarthikRIyer 
Added dispatch table to remove the conditional statements. Please review!

Also I have removed childrenIf tests for Composition as the method has not been implemented. Should we keep the childrenIf wrapper too?

KarthikRIyer commented:
@jradha11  Yes, we should keep the children if wrapper. Classes that inherit from composition should inherit that method too. If you've wrapped all the methods please mark all the eachChild and eachClip methods as deprecated with the @Deprecated annotation.

I'll take a proper look at the commits in a day or two.

jradha11 commented:
> @jradha11 Yes, we should keep the children if wrapper. Classes that inherit from composition should inherit that method too. If you've wrapped all the methods please mark all the eachChild and eachClip methods as deprecated with the @deprecated annotation.
> 
> I'll take a proper look at the commits in a day or two.

Added the @deprecated annotation.

reinecke commented on src/test/java/io/opentimeline/StackTest.java:
```diff
@@ -315,4 +316,63 @@ public void testTransformedTime() throws ChildAlreadyParentedException, NotAChil
+            assertThrows(NullPointerException.class,
+                    ()->{stack.clipIf(null, false);});
+        } catch (Exception e){
+            e.printStackTrace();
```
Why is this in a try/catch block? I would expect `assertThrows` catches the expected exception and any others would appropriately fail the test and the test framework would output the trace, right?

Right now, unexpected exceptions would still pass unit testing, right?

KarthikRIyer replied:
Good catch! I think she wanted to use try with resources to use the autocloseable functionality in Stack. But you're right. We shouldn;t use try catch here, and call close() instead.

KarthikRIyer replied:
I wonder if something like this is the reason for issue #33 🤔

KarthikRIyer replied:
We can use try-finally instead of try-catch

jradha11 replied:
@reinecke I have a few submissions this weekend so I'll have a look from tuesday and work on the tests.

reinecke reviewed (commented):
Just one question about the try/catch blocks in the tests.

KarthikRIyer commented on src/main/include/utilities.h:
```diff
@@ -1097,7 +1099,17 @@ childrenIfWrapperUtil(JNIEnv *env, jobject thisObj, jclass descendedFromCLass, j
+    jclass opt = env->GetObjectClass(searchRangeTimeRange);
+    jmethodID getMethodID = env->GetMethodID(opt, "get", "()Ljava/lang/Object;");
+    jmethodID isPresentID = env->GetMethodID(opt, "isPresent", "()Z;");
+    jboolean ifPresent = env->CallObjectMethod(searchRangeTimeRange, isPresentID);
```
isPresent() returns a boolean.
It should be `env->CallObjectMethod(searchRangeTimeRange, getMethodID);`

KarthikRIyer commented on src/main/include/utilities.h:
```diff
@@ -1097,7 +1099,17 @@ childrenIfWrapperUtil(JNIEnv *env, jobject thisObj, jclass descendedFromCLass, j
+
+    jclass opt = env->GetObjectClass(searchRangeTimeRange);
+    jmethodID getMethodID = env->GetMethodID(opt, "get", "()Ljava/lang/Object;");
+    jmethodID isPresentID = env->GetMethodID(opt, "isPresent", "()Z;");
```
The method signature would be `()Z` instead of `()Z;`
semicolons are used after classes. boolean is a primitive type.

KarthikRIyer replied:
It's not recommended to guess method signatures yourself.

You can use javap.

I did `javap -s java.util.Optional`

It gives you the signatures of all the methods in the class.

KarthikRIyer commented on src/main/include/utilities.h:
```diff
@@ -1097,7 +1099,17 @@ childrenIfWrapperUtil(JNIEnv *env, jobject thisObj, jclass descendedFromCLass, j
+    jboolean ifPresent = env->CallObjectMethod(searchRangeTimeRange, isPresentID);
+    optional<TimeRange> searchRange = nullopt;
+    if (ifPresent){
+        jobject searchRangeJobject = env->CallObjectMethod(searchRangeTimeRange, getMethodID);
```
`searchRangeJObject` instead of `searchRangeJobject`

KarthikRIyer commented on src/main/include/utilities.h:
```diff
@@ -1097,7 +1099,17 @@ childrenIfWrapperUtil(JNIEnv *env, jobject thisObj, jclass descendedFromCLass, j
     auto errorStatus = OTIO_NS::ErrorStatus();
-    optional<TimeRange> searchRange = timeRangeFromJObject(env, searchRangeTimeRange);
+
+    jclass opt = env->GetObjectClass(searchRangeTimeRange);
```
`optionalClass` instead of `opt`

KarthikRIyer commented:
Could you also remove this plugin from the build.gradle file? We do not publish to bintray.

```groovy
dependencies {
        classpath 'com.jfrog.bintray.gradle:gradle-bintray-plugin:1.+'
    }
```

KarthikRIyer commented:
Looks good.
Tiny suggestion:
```cpp
JNIEXPORT jobjectArray JNICALL Java_io_opentimeline_opentimelineio_SerializableCollection_clipIfNative
        (JNIEnv *env, jobject thisObj, jobject searchRangeTimeRange, jboolean shallowSearch)
```

I would rename `searchRangeTimeRange` to `searchRangeTimeRangeOptional` everywhere, including the utility function. It'll be very clear that the value is an optional and cannot be used directly as a timerange.

jradha11 commented:
will do the change

reinecke commented:
I think we're pretty much there! One question, I saw tests for:

```java
// Make sure we find expected clips within range
track.childrenIf(Composable.class, Optional.of(search_range), false);
```

and:

```java
// Make sure null throws NullPointerException
track.childrenIf(Composable.class, null, false);
```

But I think we were missing:
```java
// Make sure full time range is traversed
track.childrenIf(Composable.class, Optional.empty(), false);
```

jradha11 commented:
oh yes! I must have missed this edge case. Will add them too!

jradha11 commented:
Hey @reinecke @KarthikRIyer ,
I have updated tests with the trimmed and full-time range. But, I was not able to do the same for SerialisableCollectionTest. I am not getting how to filter it. Could you please review and help me out here?

reinecke commented:
Oh wow, good catch. I just checked the python implementation and it doesn't seem to have a test for this case.
Looking at [the implementation](https://github.com/PixarAnimationStudios/OpenTimelineIO/blob/8f93157e2777536c2f90a871b21edbc492698a29/src/opentimelineio/serializableCollection.h#L82), I think it is designed to:

1. ignore search range for direct children
2. apply search range for child compositions if `shallow_search` is `false`

So, with `shallow_search` as `true`, I think `search_range` has no impact on the results of  `children_if` for `SerializableCollection`. With `shallow_search` set `false`, I think it returns all the children of the `SerializableCollection` and uses type-based behavior for the children it recurses into.

Does that make sense?

Since this test is for language bindings, I don't think it needs to be too rigorous around the behavior of c++ implementation, we just want our testing to verify all the parameters are being passed into the underlying implementation correctly.

jradha11 commented:
@reinecke 
Okay! Let me work around the `shallow_search` criteria to see if filtering works. Apart from that, is there anything that is left in this PR?

KarthikRIyer commented:
@jradha11 I think what @reinecke means is that if `shallow_search = true`, `search_range` is ignored.

If `shallow_search = false`, it will take `search_range` into consideration only for Compositions (like `Track`) inside the `SerializableCollection`. `Track` organises objects temporally (one after another). So I think a `search_range` for it makes sense. Since a `Stack` organise objects spatially (one over another), but it can contain tracks, `search_range` makes sense for it too. But a `SerializableCollection` is just a collection with no specific rules. So unless it has other compositions, a `search_range` doesn't mean anything.

So you could create a `Track` with 4 clips, add it to your `SerializableCollection` and check if `clipIf` works.

@reinecke please correct me if I'm wrong.

jradha11 commented:
@KarthikRIyer @reinecke I have made the changes. I'd love if you could review! :)

reinecke commented:
Looks Good! Thanks for seeing this through!
